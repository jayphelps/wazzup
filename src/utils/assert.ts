export const assert = (value: boolean, msg = 'Assertion failed') => {
  if (value === false) {
    throw new Error(msg);
  }
};
