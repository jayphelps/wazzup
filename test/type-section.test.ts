import { ast, wasm } from '../';
import { assertModule } from './utils/assertModule';

describe('Type Section', () => {
  let module;

  beforeEach(() => {
    module = new ast.Module();
  });

  test('adds empty function types', () => {
    module.addFunctionType();
    assertModule(module, `
      | (module
      |   (type (;0;) (func)))
    `);

    module.addFunctionType();
    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (type (;1;) (func)))
    `);
  });

  test('adds function types with parameters', () => {
    module.addFunctionType([wasm.Type.i32]);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32))))
    `);

    module.addFunctionType([wasm.Type.i32, wasm.Type.i32]);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32)))
      |   (type (;1;) (func (param i32 i32))))
    `);
  });

  test('adds function types with parameters and a single return value', () => {
    module.addFunctionType([wasm.Type.i32], wasm.Type.i32);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32) (result i32))))
    `);
  });

  // wasm MVP does not yet allow multiple returns, though the binary
  // format was designed to allow it so we prepare for it
  /*
  test('adds function types with parameters and multiple return values', () => {
    module.addFunctionType([wasm.Type.i32], wasm.Type.i32);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32) (result i32))))
    `);

    module.addFunctionType([wasm.Type.i32, wasm.Type.i32], [wasm.Type.i32, wasm.Type.i32]);
    assertModule(module, `
      | (module
      |   (type (;0;) (func (param i32) (result i32))))
      |   (type (;0;) (func (param i32 i32) (result i32 i32))))
    `);
  });
  */

  test('throws if the function type has multiple return values', () => {
    expect(() => {
      module.addFunctionType([wasm.Type.i32, wasm.Type.i32], [wasm.Type.i32, wasm.Type.i32]);
    }).toThrowError('FunctionTypes can only have a single return value in wasm MVP');
  });
});
