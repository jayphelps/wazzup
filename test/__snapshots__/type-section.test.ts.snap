// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type Section adds empty function types 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Type Section adds empty function types 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  4,
  1,
  96,
  0,
  0,
]
`;

exports[`Type Section adds empty function types 3`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Type Section adds empty function types 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  7,
  2,
  96,
  0,
  0,
  96,
  0,
  0,
]
`;

exports[`Type Section adds function types with parameters 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [
        127,
      ],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Type Section adds function types with parameters 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  5,
  1,
  96,
  1,
  127,
  0,
]
`;

exports[`Type Section adds function types with parameters 3`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [
        127,
      ],
      "returnTypes": Array [],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [
        127,
        127,
      ],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Type Section adds function types with parameters 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  10,
  2,
  96,
  1,
  127,
  0,
  96,
  2,
  127,
  127,
  0,
]
`;

exports[`Type Section adds function types with parameters and a single return value 1`] = `
Module {
  "exports": Array [],
  "functions": Array [],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [
        127,
      ],
      "returnTypes": Array [
        127,
      ],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Type Section adds function types with parameters and a single return value 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  6,
  1,
  96,
  1,
  127,
  1,
  127,
]
`;
